cpu/lpc2387: implement periph/i2c

benpicco opened the pull request:
### Contribution description

This adds a driver for the I2C master interface of the lpc23xx family of SoCs.

The driver operates in interrupt mode, for every change of the internal state machine an interrupt is generated. The [data sheet](https://www.nxp.com/docs/en/user-guide/UM10211.pdf#page=505) lays out the required operations for each state change.

It doesn't seem to be possible to stop the state machine advancing without sending a stop condition, so I couldn't use the generic implementations of `i2c_read_regs()`/`i2c_write_regs()` but had to implement them in the driver using two transmit buffers. (One for the register address, one for the payload).

### Testing procedure
Unfortunately msba2 doesn't have any i2c chips connected on board.
Initially I thought sht11 was i2c, but in fact it's own bit-banging protocol.

The only i2c capable pins that are available as headers are on JP3:

  - SDA: P0.10
  - SCL: P0.11

A basic test would be to add `USEMODULE += i2c_scan` to `examples/default`.

```
2020-01-07 00:52:54,989 #  i2c_scan 0
2020-01-07 00:52:54,990 # Scanning I2C device 0...
2020-01-07 00:52:54,992 # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2020-01-07 00:52:54,993 #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2020-01-07 00:52:54,994 # 0x00 R R R R R R R R R R R R R R - -
2020-01-07 00:52:54,995 # 0x10 - - - - - - - - - - - - - - - -
2020-01-07 00:52:54,996 # 0x20 - - - - - - - - - - - - - - - -
2020-01-07 00:52:54,997 # 0x30 - - - - - - - - - - - - - - - -
2020-01-07 00:52:54,998 # 0x40 - - - - - - - - - - - - - - - -
2020-01-07 00:52:54,999 # 0x50 - - - - - - - - - - - - - - - -
2020-01-07 00:52:55,000 # 0x60 - - - - - - - - - - - - - - - -
2020-01-07 00:52:55,001 # 0x70 - - - - - - - - R R R R R R R R
```

That's still boring, so let's hook something up to it.

I used the mcb2388 with a *Tiny RTC* module (DS1307, AT24C32 - #11929) as well as a *GY-85 9-DOF* (ADXL345, ITG3200 - #10092, HMC5883 - #10083)

The chips show up in `i2c_scan`, I've successfully ran

 - `tests/driver_at24cxxx`
 - `tests/driver_adxl345`
 - `tests/driver_itg320x`

### Issues/PRs references
replacement for the driver removed in #6521
addresses #4693

maribu commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+
+    mutex_lock(&ctx[dev].tx_done);
+    return ctx[dev].res;
+}
```
Wouldn't is simplify the whole code to just use the implementations in `drivers/periph_common/i2c.c` and safe some RAM as only one buffer would be needed?

benpicco replied:
I've spend too much time trying that.
The problem was that the device's internal state machine kept advancing unless I would issue a STOP condition.

From the data sheet I've gathered that the state machine would only advance after clearing the `I2CONCLR_SIC` bit, but that didn't seem to be the case.
The device would keep generating interrupts (and reading / writing bytes) until I set `I2CONSET_STO`.

maribu replied:
OK, I see

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+{
+    switch ((uint32_t)i2c) {
+    case I2C0_BASE_ADDR:
+        PCONP |= BIT7;
```
Alright!
Turns out they were already defined in `arm7_common/include/arm7_common.h`.

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+    uint8_t buf_cur = ctx[dev].buf_cur;
+
+    /* if mode (read/write) changed, send START again */
+    if (ctx[dev].addr[buf_cur] != ctx[dev].addr[buf_cur + 1]) {
```
The function is only called if `ctx[dev].buf_cur != ctx[dev].buf_num`, that is when the last buffer has not been reached yet.

maribu replied:
If this function is indeed never called when `ctx[dev].buf_cur != 0`, there is no point in using `uint8_t buf_cur = ctx[dev].buf_cur;`. I think using `uint8_t buf_cur = 0` will safe some ROM, as e.g. addressing `ctx[dev].addr[buf_cur]` would be more efficient.

benpicco replied:
Alright!
Maybe the version before was needlessly generic - saved 20 bytes.

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+
+    ++buf_cur;
+    ctx[dev].cur = ctx[dev].buf[buf_cur];
+    ctx[dev].end = ctx[dev].buf_end[buf_cur];
```
`_next_buffer()` is only called if `ctx[dev].buf_cur != ctx[dev].buf_num`, so never for the last buffer case.

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+    case 0x48:  /* Address NACK (read)  */
+        /* send STOP */
+        i2c->CONSET = I2CONSET_STO | I2CONSET_AA;
+        _end_tx(dev, -ENXIO);
```
It's when you send e.g. a non-existing slave address.
Is the comment a bit clearer now?

maribu replied:
yep

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+    return 0;
+}
+
+static bool _install_irq(i2c_t dev)
```
I removed the return value.

benpicco commented on cpu/lpc2387/include/cpu.h:
```diff
@@ -21,6 +21,7 @@
 
 #include "lpc2387.h"
 #include "arm_cpu.h"
+#include "kernel_defines.h"
```
Yea I did use `ARRAY_SIZE` at first, but not any more.

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+ */
+#define TRX_BUFS_MAX    (2)
+static struct i2c_ctx {
+    i2c_t dev;
```
Good catch! Just a leftover…

benpicco commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,477 @@
+
+static void _set_baudrate(lpc23xx_i2c_t *i2c, uint32_t baud)
+{
+    (void) baud;
```
removed.

maribu commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,479 @@
+    uint8_t addr[TRX_BUFS_MAX];
+    uint8_t buf_num;
+    uint8_t buf_cur;
+} ctx[I2C_NUMOF];
```
How about

```C
static struct i2c_ctx {
/* ..., same as above */
} ctx[I2C_NUMOF] = {
    { .lock = MUTEX_INIT, .tx_done = MUTEX_INIT },
    { .lock = MUTEX_INIT, .tx_done = MUTEX_INIT }
};
```

That would get rid of the call to `mutex_init()`. If you take the suggestion, please squash that right in. Otherwise, just squash. I'll do one round of testing and then finally this should get in.

Sorry for the long stall. I have a lot to do recently...

benpicco replied:
This needed some additional `#ifdefs` that made is less pretty, but still saves about 200 bytes of ROM.

maribu replied:
Sounds like a good trade off :-) Please squash!

maribu replied:
Did you drop this by accident, or do you favor getting rid of the `#ifdef`s over the saved ROM size?

benpicco replied:
I think I simply forgot about it - added it now.
With `USEMODULE=i2c_scan make -C examples/default` the difference is only 48 bytes, but it's something :)

And it makes `i2c_init()` simpler, which should ease adaption to #13421

maribu commented:
OK, I tested `tests/driver_ina2xx` with the MSB-A2 and nothing connected. The expected result is that initialization fails, however, the initialization blocks and never returns:

```
2020-02-04 10:49:08,340 # main(): This is RIOT! (Version: 2020.04-devel-244-g01a49-ben)
2020-02-04 10:49:08,342 # INA2XX sensor driver test application
2020-02-04 10:49:08,342 # 
2020-02-04 10:49:08,345 # Initializing INA2XX sensor at I2C_0, address 0x00000040
```

benpicco commented:
Rebased to include #12898

The problem was that no interrupt is generated if nothing is connected to the bus.
I added a `xtimer_mutex_lock_timeout()` in b292c6800b06118bda1fe948886efbb239b7c727 that solves the issue for me.

maribu commented:
Can you drop the last fix? It turns out that this is not needed when the bug in the ina2xx driver is fixed. (*Embarrassed staring on the group*)

benpicco commented:
Alright!
There is actually a case for the timeout: I2C0 pins are always open-drain and `i2c_scan` will get stuck there if nothing is connected. (no interrupt is generated)
But I'm not sure if the added overhead is worth this, it's not required for the other i2c ports and not if something is connected to the bus.

maribu commented:
> But I'm not sure if the added overhead is worth this, it's not required for the other i2c ports and not if something is connected to the bus.

Then I'd say no :-) Can you rebase and squash? I'll give it one last test to be sure.

benpicco commented:
Rebase and squashed

benpicco commented:
@fabian18 want to give this a try too? :smiley:

fabian18 commented:
I tested `tests/driver_sht3x` with `sht31` on `msba2` and it works like a charm.
```
2020-02-26 10:37:33,640 # 
2020-02-26 10:37:33,642 # +--------Starting Measurements--------+
2020-02-26 10:37:33,655 # Temperature [°C]: 21.55
2020-02-26 10:37:33,658 # Relative Humidity [%]: 48.47
2020-02-26 10:37:33,660 # +-------------------------------------+
2020-02-26 10:37:34,663 # Temperature [°C]: 21.58
2020-02-26 10:37:34,664 # Relative Humidity [%]: 48.39
2020-02-26 10:37:34,666 # +-------------------------------------+
2020-02-26 10:37:35,654 # Temperature [°C]: 21.58
2020-02-26 10:37:35,669 # Relative Humidity [%]: 48.35
2020-02-26 10:37:35,670 # +-------------------------------------+
2020-02-26 10:37:36,681 # Temperature [°C]: 21.61
2020-02-26 10:37:36,682 # Relative Humidity [%]: 48.31
2020-02-26 10:37:36,684 # +-------------------------------------+
```

benpicco commented:
rebased to fix missing header in `at24mac`.

gschorcht commented on cpu/lpc2387/periph/i2c.c:
```diff
@@ -0,0 +1,484 @@
+    case 0x28: /* Data byte has been transmitted */
+
+        if (ctx[dev].cur == ctx[dev].end) {
+            /* we transmited all buffers */
```
```suggestion
            /* we transmitted all buffers */
```

gschorcht commented:
Please squash.

benpicco commented:
squashed.

benpicco commented:
Thank you all for reviewing, testing and merging 😄
